js/shared/core.js

```diff
@@ -21,7 +21,103 @@ function ocaml_terminal_size_get() {
 //Provides: unix_open
 //Requires: caml_sys_open
 function unix_open(path, flags, perm) {
-  return caml_sys_open(path, flags, perm);
+  console.log("unix_open");
+  console.log(path);
+  console.log(flags);
+  console.log(perm);
+
+  // Why is this needed?
+  // There is a Unix.open_flag type, which is different than
+  // that of the Stdlib.open_flag type.
+  // As such, some flags (such as O_CREAT) appear in a different
+  // index of the variant type than the corresponding Open_creat,
+  // for instance.
+  // So before we can call the system open command, we need to
+  // translate the Unix open flags to stdlib open flags.
+  function map_unix_flag_to_stdlib_flag(flag) {
+    switch (flag) {
+      // O_RDONLY
+      case 0:
+        return [0];
+      // O_WRONLY
+      case 1:
+        return [1];
+      // O_RDWR
+      case 2:
+        return [0, 1];
+      //O_NONBLOCK
+      case 3:
+        return [8];
+      //O_APPEND
+      case 4:
+        return [2];
+      //O_CREAT
+      case 5:
+        return [3];
+      //O_TRUNC
+      case 6:
+        return [4];
+      //O_EXCL
+      case 7:
+        return [5];
+      //O_NOCTTY
+      case 8:
+        // TODO
+        return [];
+      //O_DSYNC
+      case 9:
+        // TODO
+        return [];
+      //O_SYNC
+      case 10:
+        // TODO
+        return [];
+      //O_RSYNC
+      case 11:
+        // TODO
+        return [];
+      //O_SHARE_DELETE
+      case 12:
+        // TODO
+        return [];
+      //O_CLOEXEC
+      case 13:
+        // TODO
+        return [];
+      //O_KEEPEXEC
+      case 14:
+        // TODO
+        return [];
+    }
+  }
+
+  function parse_flags(flags) {
+    if (flags == 0) {
+      return [];
+    } else {
+      let elem = flags[1];
+      return [elem].concat(parse_flags(flags[2]));
+    }
+  }
+
+  function unparse_flags(flags) {
+    if (flags.length == 0) {
+      return 0;
+    } else {
+      return [0, flags[0], unparse_flags(flags.slice(1))];
+    }
+  }
```

Comment:
are these methods basically just translating between ocaml and js list representations? are there jsoo runtime methods we could use instead?

Reply:
(`caml_list_of_js_array` / `caml_list_to_js_array`)